presto-kafka/src/test/java/io/prestosql/plugin/kafka/encoder/json/TestMillisecondsJsonDateTimeFormatter.java

```diff
@@ -0,0 +1,87 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka.encoder.json;
+
+import io.prestosql.plugin.kafka.encoder.json.format.DateTimeFormat;
+import io.prestosql.spi.type.SqlTime;
+import io.prestosql.spi.type.SqlTimeWithTimeZone;
+import io.prestosql.spi.type.SqlTimestamp;
+import io.prestosql.spi.type.SqlTimestampWithTimeZone;
+import io.prestosql.spi.type.TimeZoneKey;
+import org.testng.annotations.Test;
+
+import java.time.ZoneOffset;
+import java.util.Optional;
+
+import static io.prestosql.plugin.kafka.encoder.json.format.util.TimeConversions.scalePicosToMillis;
+import static io.prestosql.spi.type.TimeZoneKey.UTC_KEY;
+import static io.prestosql.testing.DateTimeTestingUtils.sqlTimeOf;
+import static io.prestosql.testing.DateTimeTestingUtils.sqlTimeWithTimeZoneOf;
+import static io.prestosql.testing.DateTimeTestingUtils.sqlTimestampOf;
+import static io.prestosql.testing.DateTimeTestingUtils.sqlTimestampWithTimeZoneOf;
+import static io.prestosql.testing.assertions.Assert.assertEquals;
+
+public class TestMillisecondsJsonDateTimeFormatter
+{
+    private void testTime(SqlTime value, DateTimeFormat dateTimeFormat, int precision)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.empty()).formatTime(value, precision);
+        assertEquals(Long.parseLong(formattedStr), scalePicosToMillis(value.getPicos()));
+    }
+
+    private void testTimeWithTZ(SqlTimeWithTimeZone value, DateTimeFormat dateTimeFormat)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.empty()).formatTimeWithZone(value);
+        assertEquals(Long.parseLong(formattedStr), value.getMillisUtc());
+    }
+
+    private void testTimestamp(SqlTimestamp value, DateTimeFormat dateTimeFormat)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.empty()).formatTimestamp(value);
+        assertEquals(Long.parseLong(formattedStr), value.getMillis());
+    }
+
+    private void testTimestampWithTZ(SqlTimestampWithTimeZone value, DateTimeFormat dateTimeFormat)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.empty()).formatTimestampWithZone(value);
+        assertEquals(Long.parseLong(formattedStr), value.getEpochMillis());
+    }
+
+    @Test
+    public void testMillisecondsDateTimeFunctions()
+    {
+        DateTimeFormat dateTimeFormat = DateTimeFormat.MILLISECONDS_SINCE_EPOCH;
+
+        testTime(sqlTimeOf(3, 15, 36, 25, 123000000), dateTimeFormat, 3);
```

Comment:
explicitly pass expected values. To avoid hard to read seconds-since-epoch you can have some trivial generation logic, but I would suggest to do it in place. E.g sth like:
```
testTime(sqlTimeOf(3, 15, 36, 25, 123000000), dateTimeFormat, 3, LocalTime.of(15, 36, 25, 123000000).toNanoOfDay()/1000);
```